rust/onnxruntime/src/session.rs

```diff
@@ -0,0 +1,742 @@
+    pub outputs: Vec<Output>,
+}
+
+/// Information about an ONNX's input as stored in loaded file
+#[derive(Debug)]
+pub struct Input {
+    /// Name of the input layer
+    pub name: String,
+    /// Type of the input layer's elements
+    pub input_type: TensorElementDataType,
+    /// Shape of the input layer
+    ///
+    /// C API uses a i64 for the dimensions. We use an unsigned of the same range of the positive values.
+    pub dimensions: Vec<Option<u32>>,
+}
+
+/// Information about an ONNX's output as stored in loaded file
+#[derive(Debug)]
+pub struct Output {
+    /// Name of the output layer
+    pub name: String,
+    /// Type of the output layer's elements
+    pub output_type: TensorElementDataType,
+    /// Shape of the output layer
+    ///
+    /// C API uses a i64 for the dimensions. We use an unsigned of the same range of the positive values.
+    pub dimensions: Vec<Option<u32>>,
+}
+
+impl Input {
+    /// Return an iterator over the shape elements of the input layer
+    ///
+    /// Note: The member [`Input::dimensions`](struct.Input.html#structfield.dimensions)
+    /// stores `u32` (since ONNX uses `i64` but which cannot be negative) so the
+    /// iterator converts to `usize`.
+    pub fn dimensions(&self) -> impl Iterator<Item = Option<usize>> + '_ {
+        self.dimensions.iter().map(|d| d.map(|d2| d2 as usize))
+    }
+}
+
+impl Output {
+    /// Return an iterator over the shape elements of the output layer
+    ///
+    /// Note: The member [`Output::dimensions`](struct.Output.html#structfield.dimensions)
+    /// stores `u32` (since ONNX uses `i64` but which cannot be negative) so the
+    /// iterator converts to `usize`.
+    pub fn dimensions(&self) -> impl Iterator<Item = Option<usize>> + '_ {
+        self.dimensions.iter().map(|d| d.map(|d2| d2 as usize))
+    }
+}
+
+impl Drop for Session {
+    #[tracing::instrument]
+    fn drop(&mut self) {
+        debug!("Dropping the session.");
+        if self.session_ptr.is_null() {
+            error!("Session pointer is null, not dropping.");
+        } else {
+            unsafe { g_ort().ReleaseSession.unwrap()(self.session_ptr) };
+        }
+
+        self.session_ptr = std::ptr::null_mut();
+        self.allocator_ptr = std::ptr::null_mut();
+    }
+}
+
+unsafe impl Send for Session {}
+
+impl Session {
+    /// Run the input data through the ONNX graph, performing inference.
+    ///
+    /// Note that ONNX models can have multiple inputs; a `Vec<_>` is thus
+    /// used for the input data here.
+    pub fn run<'a>(
+        &self,
+        input_arrays: Vec<&mut dyn ConstructTensor>,
+    ) -> Result<Vec<OrtOutput<'a>>> {
+        let memory_info = &self.memory_info;
+
+        let allocator = self.allocator_ptr;
+
+        let input_arrays = input_arrays
+            .into_iter()
+            .map(|v| v.construct(memory_info, allocator))
+            .collect::<Result<Vec<_>>>()?;
```

Comment:
```suggestion
    pub fn run<'a>(
        &self,
        mut input_arrays: impl AsMut<[&'a mut dyn ConstructTensor]>,
    ) -> Result<Vec<OrtOutput<'a>>> {
        let memory_info = &self.memory_info;

        let allocator = self.allocator_ptr;

        let input_arrays = input_arrays
            .as_mut()
            .iter_mut()
            .map(|v| v.construct(memory_info, allocator))
            .collect::<Result<Vec<_>>>()?;
```

I want run function to be able to receive any array like types.
`impl AsMut<[&'a mut dyn ConstructTensor]>` can receive `Vec<&'a mut dyn ConstructTensor>` and `[&'a mut dyn ConstructTensor]` types.

However, It has cons. Because run function's API becomes difficult to understand.

Reply:
I like this idea.

Reply:
Check out if what I did was what you were thinking.

Reply:
@boydjohnson Can you tell me why use Box to input_arrays?

Reply:
Hi @qwerty2501. I chose Box<dyn ConstructTensor> because it is owned and the original Session::run signature had owned values. I want folks who have code based on the original bindings won't have to do much to update to these bindings.

Reply:
Thank you. I see.